### PR TITLE
Duet AI code-explanation and code-generation prompts basic

### PR DESCRIPTION
**Pull request by olehborysevych**
Duet AI prompts for:
CSV
JSON
Parquet
Avro
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        }));
+
+        p.run();
+    }
+}
+```
+
+Response:
+
+This code reads data from [Apache Avro](https://avro.apache.org/) files using the `ReadFromAvro` transform from the built-in [AvroIO] (https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/extensions/avro/io/AvroIO.html) connector.
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:45]:
```suggestion
This code reads data from an Apache Avro file using the `ReadFromAvro` transform from the AvroIO connector provided by Apache Beam.
```

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+```java
+    public interface ReadAvroOptions extends PipelineOptions {
+        @Description("Path to the Avro file to read from")
+        @Default.String("gs://<your-bucket-name]/[your-object-name]/filename.avro")
+        String getInput();
+
+        void setInput(String value);
+    }
+```
+The `ReadAvroOptions` interface is used to define the command-line argument `--input`, which specifies the path to the Avro file. Set default value using the `@Default` annotation.
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:56]:
```suggestion

The `ReadAvroOptions` interface defines the command-line argument `--input`, which specifies the path to the Avro file and sets the default value using the `@Default` annotation.
```

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+```java
+        ReadAvroOptions options = PipelineOptionsFactory.fromArgs(args).withValidation().as(ReadAvroOptions.class);
+
+        Pipeline p = Pipeline.create(options);
+```
+`PipelineOptionsFactory` ia used to parse command-line arguments and create a ReadAvroOptions instance. This instance is used to create a pipeline with provided pipeline options.
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:63]:
```suggestion

The `PipelineOptionsFactory` class parses the command-line arguments and creates a `ReadAvroOptions` instance. This instance is then used to create a pipeline with the provided pipeline options.
```

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+```java
+        String fields = "[{ \"name\": \"months\", \"type\": \"string\" }, { \"name\": \"amount\", \"type\": \"string\" }]";
+        String schema = "{\"type\": \"record\", \"name\": \"AvroRecord\",\"fields\":" + fields + "}";
+        Schema avroSchema = Schema.parse(schema);
+```
+AvroIO connector requires a schema to read Avro files. The schema is defined as a string and parsed into a `Schema` object.
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:70]:
```suggestion

The AvroIO connector requires a schema to read Avro files. Hence, the schema is defined as a string and parsed into a `Schema` object.
```

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            public void processElement(ProcessContext c) {
+                GenericRecord record = c.element();
+                String message = "Month: " + record.get("month") + ". Amount: " + record.get("amount");
+                LOG.info(message);
+                c.output(message);
+            }
+        }))
+```
+
+`ParDo` transform is used to process each `GenericRecord` object from the Avro file. Parse the `GenericRecord` object into a string and log it.
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:85]:
```suggestion
In this segment, the `ParDo` transform processes each `GenericRecord` object from the Avro file. Each `GenericRecord` object is then parsed into a string and logged accordingly.
```

File: learning/prompts/code-explanation/java/09_io_avro.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+```java
+        p.run();
+```
+Pipeline is executed to read the Avro file using the AvroIO connector, parse the `GenericRecord` object, format it and output the result.
+
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/09_io_avro.md:91]:
```suggestion

Finally, the pipeline is executed to read the Avro file using the AvroIO connector, parse the `GenericRecord` objects, format them, and output the results.
```

**Review by dariabezkorovaina — CHANGES_REQUESTED** [review]
@olehborysevych, looks great, minor edits for grammar and consistency:) Also, removed all the links. Please change the directory structure to "code-explanation-nolinks" and code-generation-nolinks", so that the prompts do not mix up with the previous prompts with links.

File: learning/prompts/code-explanation/java/08_io_csv.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            .apply(
+                "Write to CSV file",
+                    CsvIO.<ExampleRecord>write(options.getFilePathPrefix(), csvFormat)
+                        .withNumShards(1));
+        p.run();
+    }
+}
+
+```
+Response:
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/08_io_csv.md:59]:
```suggestion

Response:
```

File: learning/prompts/code-explanation/java/10_io_parquet.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                                record.get(ExampleRecord.COLUMN_ID),
+                                record.get(ExampleRecord.COLUMN_MONTH),
+                                record.get(ExampleRecord.COLUMN_AMOUNT));
+                            c.output(record);
+                        }
+                    }));
+        p.run();
+    }
+}
+
```
Comment (dariabezkorovaina) [inline learning/prompts/code-explanation/java/10_io_parquet.md:77]:
```suggestion
```

**Review by dariabezkorovaina — APPROVED** [review]
Looks good, 2 nits removing/adding an extra empty line :)

**Comment by olehborysevych** [comment]
Hey @damccorm this one is ready for review as well

**Review by damccorm — APPROVED** [review]
Thanks!

**Pull request merged**